features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/domain/RequestBodyFactory.kt

```diff
@@ -18,78 +19,45 @@ internal class RequestBodyFactory(
     private val compressor: BytesCompressor = BytesCompressor()
 ) {
 
-    fun create(
-        segment: MobileSegment,
-        serializedSegment: JsonObject
-    ): RequestBody {
-        // we need to add a new line at the end of each segment for being able to format it
-        // as an Array when read by the player
-        val segmentAsBinary = (serializedSegment.toString() + "\n").toByteArray()
-        return buildRequestBody(segment, segmentAsBinary)
+    fun create(serializedSegmentsPairs: List<Pair<MobileSegment, JsonObject>>): RequestBody {
+        return buildRequestBody(serializedSegmentsPairs)
     }
 
-    @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
-    private fun buildRequestBody(segment: MobileSegment, segmentAsBinary: ByteArray): RequestBody {
-        val compressedData = compressor.compressBytes(segmentAsBinary)
-        return MultipartBody.Builder()
-            .setType(MultipartBody.FORM)
-            .addFormDataPart(
-                SEGMENT_FORM_KEY,
-                segment.session.id,
-                compressedData
-                    .toRequestBody(CONTENT_TYPE_BINARY.toMediaTypeOrNull())
-            )
-            .addFormDataPart(
-                APPLICATION_ID_FORM_KEY,
-                segment.application.id
-            )
-            .addFormDataPart(
-                SESSION_ID_FORM_KEY,
-                segment.session.id
-            )
-            .addFormDataPart(
-                VIEW_ID_FORM_KEY,
-                segment.view.id
-            )
-            .addFormDataPart(
-                HAS_FULL_SNAPSHOT_FORM_KEY,
-                segment.hasFullSnapshot.toString()
-            )
-            .addFormDataPart(
-                RECORDS_COUNT_FORM_KEY,
-                segment.recordsCount.toString()
-            )
-            .addFormDataPart(
-                RAW_SEGMENT_SIZE_FORM_KEY,
-                compressedData.size.toString()
-            )
-            .addFormDataPart(
-                START_TIMESTAMP_FORM_KEY,
-                segment.start.toString()
-            )
-            .addFormDataPart(
-                END_TIMESTAMP_FORM_KEY,
-                segment.end.toString()
-            )
-            .addFormDataPart(
-                SOURCE_FORM_KEY,
-                segment.source.toJson().asString
-            )
-            .build()
+    @Suppress("UnsafeThirdPartyFunctionCall")
+    private fun buildRequestBody(segments: List<Pair<MobileSegment, JsonObject>>): RequestBody {
+        val multipartBody = MultipartBody.Builder().setType(MultipartBody.FORM)
+        val metadata = JsonArray()
+        segments.forEachIndexed { index, segment ->
+            val segmentAsByteArray = (segment.second.toString() + "\n").toByteArray()
```

Comment:
why do we need to add `\n` here?

Reply:
well...it is a problem with the way the payloads are being concatenated when decompressed so they need this extra new line. I am going to add a comment there not to forget about it.